### PR TITLE
Modify packages for the user instead of printing instructions when adding package.

### LINKED ISSUE #17902: Modify packages for user instead of printing instructions when adding packages
When running `pulumi package add`, add the dependency directly instead of printing instructions.

### PR DESCRIPTION
**Pull request by brandonpollack23**
Before we would print instructions, but with built in tools it is not
hard to modify the go.mod on behalf of the user and print out
information.

This still leaves java in a "print instruction" state since there isnt a
standard package managment system like in the other languages.

Fixes #17902

### CODE REVIEW

File: pkg/cmd/pulumi/package_add.go [hunk 1]
```diff
@@ -204,18 +207,50 @@ func printPythonLinkInstructions(ws pkgWorkspace.Context, root string, pkg *sche
 	if err != nil {
 		return err
 	}
-	pipInstructions := func() {
-		fmt.Printf("  echo %s >> requirements.txt\n\n", packageSpecifier)
-		fmt.Printf("  pulumi install\n")
+
+	modifyRequirments := func() error {
```
Comment (tgummerer) [inline pkg/cmd/pulumi/package_add.go:211]:
```suggestion
	modifyRequirements := func() error {
```

File: pkg/cmd/pulumi/package_add.go [hunk 2]
```diff
@@ -204,18 +207,50 @@ func printPythonLinkInstructions(ws pkgWorkspace.Context, root string, pkg *sche
...
+		}
+
+		return nil
 	}
+
 	options := proj.Runtime.Options()
 	if toolchain, ok := options["toolchain"]; ok {
 		if tc, ok := toolchain.(string); ok {
 			switch tc {
 			case "pip":
```
Comment (tgummerer) [inline pkg/cmd/pulumi/package_add.go:242]: Not new in your PR, but we're missing `uv` here.

**Review by tgummerer — COMMENTED** [review]
We should have tests for this.

**Comment by Frassle** [comment]
> We should have tests for this.

Aye, I like the ideas here but I don't want this merged till we're testing this in conformance.

**Comment by brandonpollack23** [comment]
> > We should have tests for this.
> 
> Aye, I like the ideas here but I don't want this merged till we're testing this in conformance.

My understanding of conformance tests is they test a set up project and compare it's snapshot with expected states etc.  This is more of an integration test isn't it?  So we should just do a regular test here to test that the `pulumi package add` command itself generates an SDK and modifies the project file as expected, right?

**Comment by Frassle** [comment]
> My understanding of conformance tests is they test a set up project and compare it's snapshot with expected states etc

Yes, but I do want conformance testing both "published" and "local" sdks, and checking the language hosts can deal with both.

> So we should just do a regular test here to test that the pulumi package add command itself generates an SDK and modifies the project file as expected, right?

That would be ok to do if much easier than doing it in conformance (probably true). But I'd hope to clean those integrations tests up to just be conformance tests eventually.

**Comment by brandonpollack23** [comment]
> > My understanding of conformance tests is they test a set up project and compare it's snapshot with expected states etc
> 
> Yes, but I do want conformance testing both "published" and "local" sdks, and checking the language hosts can deal with both.
> 
> > So we should just do a regular test here to test that the pulumi package add command itself generates an SDK and modifies the project file as expected, right?
> 
> That would be ok to do if much easier than doing it in conformance (probably true). But I'd hope to clean those integrations tests up to just be conformance tests eventually.

I added integration tests for now

File: pkg/cmd/pulumi/package_add.go [hunk 3]
```diff
@@ -291,18 +352,22 @@ func csharpPackageName(pkgName string) string {
...
+	cmd := exec.Command("dotnet", "add", "reference", relOut)
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stdout
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("dotnet error: %w", err)
+	}
+
 	fmt.Printf("You also need to add the following to your .csproj file of the program:\n")
 	fmt.Println()
 	fmt.Println("  <DefaultItemExcludes>$(DefaultItemExcludes);sdks/**/*.cs</DefaultItemExcludes>")
```
Comment (Frassle) [inline pkg/cmd/pulumi/package_add.go:373]: Can't we do this part automatically as well?

**Comment by julienp** [comment]
I am not opposed to doing it in the CLI for now and it is a great improvement over just printing instructions, but I thought the plan was to move this down into the language runtimes eventually. The language runtimes know best how to deal with their ecosystem, and have the full knowledge about which package manager is in use. For example right now for nodejs we only look at the packagemanager option, but `install` (via the language runtime) knows how to detect the packagemanager based on the presence of lock files or env vars too.

File: pkg/cmd/pulumi/package_add.go (hunk 3 above)
Comment (brandonpollack23) [reply pkg/cmd/pulumi/package_add.go:373]: I can try, I was short on time and needed to get back to main work for now, Mind if i just file a followup issue?

**Comment by brandonpollack23** [comment]
Yay then this is finally ready

File: pkg/cmd/pulumi/package_add.go (hunk 3 above)
Comment (Frassle) [reply pkg/cmd/pulumi/package_add.go:373]: Given the dotnet stuff isn't tested yet just back out the changes for this, we can leave it as a "just print" for the time being like java.

**Comment by brandonpollack23** [comment]
So this never merged, I'm rebasing and trying again...

File: pkg/cmd/pulumi/package_add.go (hunk 3 above)
Comment (Frassle) [reply pkg/cmd/pulumi/package_add.go:373]: This comment still stands, I don't think we should try changing half of dotnet, especially without tests. Just leave it as is for now.

**Pull request merged**